gax/src/main/java/com/google/api/gax/core/Distribution.java

```diff
@@ -42,117 +42,60 @@
+    // NOTE: This implementation uses the nearest-rank method.
+    // https://en.wikipedia.org/wiki/Percentile#The_nearest-rank_method
+    //
+    // If called concurrently with record, this implementation biases low.
+    // Since we read count before iterating the array, more elements might have been recorded
+    // and we might not iterate far enough to be fair.
+    // Still, this probably won't matter greatly in practice.
+
     Preconditions.checkArgument(percentile > 0.0);
     Preconditions.checkArgument(percentile <= 100.0);
 
-    long[] bucketCounts = getBucketCounts();
-    long total = 0;
-    for (long count : bucketCounts) {
-      total += count;
-    }
-
-    if (total == 0) {
-      return 0;
-    }
-    long count = (long) Math.ceil(total * percentile / 100.0);
-    for (int i = 0; i < bucketCounts.length; i++) {
-      count -= bucketCounts[i];
-      if (count <= 0) {
+    long targetRank = (long) Math.ceil(percentile * count.get() / 100);
+    long rank = 0;
+    for (int i = 0; i < buckets.length(); i++) {
+      rank += buckets.get(i);
+      if (rank >= targetRank) {
         return i;
       }
     }
-    return 0;
-  }
-
-  /** Resets (sets to 0) the recorded values. */
-  public synchronized void reset() {
-    for (AtomicLong element : bucketCounts) {
-      element.set(0);
-    }
-    count = 0;
-    mean = 0;
-    sumOfSquaredDeviation = 0;
-  }
-
-  /** Numbers of values recorded. */
-  public long getCount() {
-    return count;
-  }
-
-  /** Square deviations of the recorded values. */
-  public double getSumOfSquareDeviations() {
-    return sumOfSquaredDeviation;
-  }
-
-  /** Mean of the recorded values. */
-  public double getMean() {
-    return mean;
-  }
-
-  /** Gets the accumulated count of every bucket of the distribution. */
-  public long[] getBucketCounts() {
-    long[] counts = new long[bucketCounts.length];
-    for (int i = 0; i < counts.length; i++) {
-      counts[i] = bucketCounts[i].longValue();
-    }
-    return counts;
-  }
-
-  /** Make a copy of the distribution. */
-  public synchronized Distribution copy() {
-    Distribution distributionCopy = new Distribution(bucketCounts.length);
-    distributionCopy.count = count;
-    distributionCopy.mean = mean;
-    distributionCopy.sumOfSquaredDeviation = sumOfSquaredDeviation;
-    for (int i = 0; i < bucketCounts.length; i++) {
-      distributionCopy.bucketCounts[i].set(bucketCounts[i].get());
-    }
-    return distributionCopy;
+    return buckets.length();
   }
 
   /** Record a new value. */
   public void record(int bucket) {
     Preconditions.checkArgument(bucket >= 0);
 
-    synchronized (this) {
-      count++;
-      double dev = bucket - mean;
-      mean += dev / count;
-      sumOfSquaredDeviation += dev * (bucket - mean);
+    // Account for bucket overflow, records everything that is equals or greater of the last
+    // bucket.
+    if (bucket >= buckets.length()) {
+      bucket = buckets.length() - 1;
```

Comment:
There is some stuff going on here that is not documented on the surface and it's quite confusing. Basically, it looks like the value of `bucket` is capped at `totalBuckets - 1`, right? What is the meaning of "bucket"?

Reply:
Fixed docs. The "buckets" are an implementation detail. If your bucket is `[0,5,2,8]` we have record 5 ones, 2 twos. and 8 threes, etc.